# SVM baseline: three training variants

Fits an RBF SVM on the **same partitions the CNN used**, reading
`cnn-latest/results/splits/{variant}-seed{seed}.csv`. `SVC` is deterministic given its
data, so the split is the only source of run-to-run variation and every SVM checkpoint is
*paired* with a CNN checkpoint rather than merely matched in protocol.

| variant | pool | partition |
| --- | --- | --- |
| `orig` | `training.npz` (7,200) | `orig-seed{s}` |
| `clean` | `training-clean.npz` (7,172) | `clean-seed{s}` |
| `clean-fixed` | `training.npz` minus the 28 defective rows | `orig-seed{s}`, rows dropped in place |

`orig` vs `clean` changes the data *and* redraws the split. `clean-fixed` changes only the
data, so it is the arm that isolates the 28 clips as a cause.

`C=1000, gamma=1e-5` are frozen across every fit, inherited from the published tuning, so
the variant comparison is an ablation and not a re-tuning exercise. Fitting uses the
`train` split only, never `train+val`, so the SVM sees exactly the rows the CNN trained
on — which means these runs will *not* reproduce `models/svm-best`, fit on a 90/10 split.

Each fit runs in its own OS process. That is not stylistic: a fitted model holds a 1.35 GB
float64 support-vector array, and process exit is the only thing that reliably returns it.

In [1]:
import subprocess, sys, os, time
from pathlib import Path
import numpy as np, pandas as pd

HERE = Path("/home/seya/code/chord-detection/training/notebooks/svm-latest")
PY = "/home/seya/code/chord-detection/.venv/bin/python"
sys.path.insert(0, str(HERE))
import svm_common as C

C.ensure_dirs()
print("training root:", C.TRAINING_ROOT)
print("variants:", C.VARIANTS)
print("seeds:", C.SEEDS)
print("frozen hyperparameters: C=%g gamma=%g" % (C.DEFAULT_C, C.DEFAULT_GAMMA))

rows = []
for v in C.VARIANTS:
    for tag in ["seed%d" % s for s in C.SEEDS]:
        idx = C.split_indices(C.load_split(v, tag))
        rows.append({"variant": v, "tag": tag,
                     **{k: len(x) for k, x in idx.items()}})
pd.DataFrame(rows).pivot_table(index="variant", columns="tag",
                               values=["train", "val", "test"], aggfunc="first")

training root: /home/seya/code/chord-detection/training
variants: ['orig', 'clean', 'clean-fixed']
seeds: [42, 43, 44, 45, 46]
frozen hyperparameters: C=1000 gamma=1e-05


test                              train                       \
tag         seed42 seed43 seed44 seed45 seed46 seed42 seed43 seed44 seed45   
variant                                                                      
clean          718    718    718    718    718   5736   5736   5736   5736   
clean-fixed    718    717    719    719    716   5736   5736   5736   5741   
orig           720    720    720    720    720   5760   5760   5760   5760   

                      val                              
tag         seed46 seed42 seed43 seed44 seed45 seed46  
variant                                                
clean         5736    718    718    718    718    718  
clean-fixed   5739    718    719    717    712    717  
orig          5760    720    720    720    720    720

## Step zero — prove the fast predict is exact

`SVC.predict` runs single-threaded libsvm, which streams the whole 1.35 GB float64
support-vector array once per clip. At 0.27–0.43 s/clip the planned campaign is ~20 hours
of prediction alone, so `fast_rbf.FastRBF` recomputes the same decision values with one
float32 GEMM per chunk plus 36 small per-class matmuls.

Being *fast* is worthless if it is not *identical*. This cell checks the reconstruction
against the published `models/svm-best` on `vivo` — the hardest available rows, whose
scaled norms are ~1.7× the training set's, so kernel values are smallest and decision
values are closest to a sign flip. It raises on any single disagreement.

In [5]:
r = subprocess.run([PY, str(HERE / "gate_svmbest.py"), "--n", "96"],
                   capture_output=True, text=True, cwd=HERE)
print(r.stdout or r.stderr[-3000:])
assert "GATE PASSED" in r.stdout, "fast predict does not match libsvm - stop here"

svm-best: 36 classes, 4143 SVs, gamma 1e-05, C 1000
libsvm         185 ms/clip
mean max K     0.5084

f32_gemm
  agreement    96/96 exact
  dec error    max 2.364e-07  median 9.546e-09
  min |dec|    8.055e-06  -> margin 34.1x
  speed        1.00 ms/clip  (185x libsvm)

f64_gemm
  agreement    96/96 exact
  dec error    max 1.721e-14  median 3.279e-15
  min |dec|    8.055e-06  -> margin 4.68e+08x
  speed        1.97 ms/clip  (94x libsvm)

GATE PASSED



## First fit — reproduce the known result before trusting anything downstream

`orig` seed 42 should land near the previously recorded `thinkpad-2` accuracy of 0.572
with `C_diminished_4` absorbing ~37% of predictions. It will not match exactly: the
published `svm-best` was fit on a 90/10 split of the whole pool (6,480 rows), while this
fits the CNN's `train` split (5,760 rows). Close is the bar; far means the pipeline
differs from the original run and nothing after this point is trustworthy.

In [6]:
def fit_one(variant, seed, extra=()):
    """One (variant, seed) in its own process; a 1.35 GB SV array is only reclaimed on exit."""
    t = time.time()
    cmd = [PY, str(HERE / "run_one.py"), "--variant", variant, "--seed", str(seed), *extra]
    p = subprocess.Popen(cmd, cwd=HERE, stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT, text=True, bufsize=1)
    for line in p.stdout:
        print(line.rstrip(), flush=True)
    p.wait()
    print(f"[{variant}/seed{seed} exit {p.returncode} in {time.time() - t:.0f}s]")
    return p.returncode

fit_one("orig", 42)

orig/seed42: train 5760 val 720 test 720


  fit 316.3s  nSV 3893


  gate OK 128 rows, min|dec| 7.569e-06 (30s)


  thinkpad    acc 0.8597  sink C_diminished_4 14.9%


  vivo        acc 0.9750  sink G_minor_4 4.0%


  flow        acc 0.9938  sink C_diminished_4 3.1%


  thinkpad-2  acc 0.5569  sink C_diminished_4 38.0%


  flow-2      acc 0.7743  sink C_diminished_4 18.1%


OK orig/seed42  test 1.0000  1.89 ms/clip


[orig/seed42 exit 0 in 423s]


0

## The full sweep

Three variants × five splits. `run_one.py` skips any `(variant, seed)` already in
`progress.csv`, so re-running this cell resumes rather than repeats — `orig` seed 42 above
is already done and will be skipped.

Runs are strictly sequential. Two concurrent fits each hold a multi-GB support-vector
array, and a measured attempt at that degraded libsvm by ~20× through memory-bandwidth
contention alone. Roughly 7 minutes per run.

In [ ]:
t0 = time.time()
for variant in C.VARIANTS:
    for seed in C.SEEDS:
        fit_one(variant, seed)
print(f"\nsweep finished in {(time.time() - t0) / 60:.0f} min")

skip orig/seed42 (already in progress.csv)


[orig/seed42 exit 0 in 1s]


orig/seed43: train 5760 val 720 test 720


  fit 290.9s  nSV 3865


  gate OK 128 rows, min|dec| 2.547e-06 (31s)


  thinkpad    acc 0.8806  sink C_diminished_4 13.8%


  vivo        acc 0.9778  sink G_minor_4 4.0%


  flow        acc 0.9951  sink C_diminished_4 3.1%


  thinkpad-2  acc 0.5778  sink C_diminished_4 37.0%


  flow-2      acc 0.8021  sink C_diminished_4 15.4%


OK orig/seed43  test 1.0000  1.48 ms/clip


[orig/seed43 exit 0 in 372s]


orig/seed44: train 5760 val 720 test 720


  fit 244.1s  nSV 3861


  gate OK 128 rows, min|dec| 3.054e-07 (23s)


  thinkpad    acc 0.8056  sink C_diminished_4 18.0%


  vivo        acc 0.9743  sink E_minor_4 4.0%


  flow        acc 0.9938  sink C_diminished_4 3.3%


  thinkpad-2  acc 0.5319  sink C_diminished_4 42.1%


  flow-2      acc 0.7451  sink C_diminished_4 19.9%


OK orig/seed44  test 1.0000  1.62 ms/clip


[orig/seed44 exit 0 in 313s]


orig/seed45: train 5760 val 720 test 720


  fit 268.3s  nSV 3913


  gate OK 128 rows, min|dec| 4.834e-06 (25s)


  thinkpad    acc 0.7701  sink C_diminished_4 22.6%


  vivo        acc 0.9750  sink G_minor_4 4.5%


  flow        acc 0.9903  sink C_diminished_4 3.5%


  thinkpad-2  acc 0.4625  sink C_diminished_4 48.6%


  flow-2      acc 0.7118  sink C_diminished_4 24.9%


OK orig/seed45  test 1.0000  1.49 ms/clip


[orig/seed45 exit 0 in 334s]


orig/seed46: train 5760 val 720 test 720


  fit 227.5s  nSV 3929


  gate OK 128 rows, min|dec| 1.666e-06 (25s)


  thinkpad    acc 0.8097  sink C_diminished_4 19.8%


  vivo        acc 0.9757  sink G_minor_4 4.1%


  flow        acc 0.9903  sink C_diminished_4 3.6%


  thinkpad-2  acc 0.4743  sink C_diminished_4 47.9%


  flow-2      acc 0.7542  sink C_diminished_4 22.6%


OK orig/seed46  test 1.0000  1.46 ms/clip


[orig/seed46 exit 0 in 293s]


clean/seed42: train 5736 val 718 test 718


  fit 234.1s  nSV 4037


  gate OK 128 rows, min|dec| 1.652e-05 (25s)


  thinkpad    acc 0.7889  sink C_diminished_4 20.8%


  vivo        acc 0.9701  sink E_minor_4 4.3%


  flow        acc 0.9931  sink C_diminished_4 3.1%


  thinkpad-2  acc 0.4271  sink C_diminished_4 46.7%


  flow-2      acc 0.6979  sink C_diminished_4 19.2%


OK clean/seed42  test 1.0000  1.64 ms/clip


[clean/seed42 exit 0 in 309s]


clean/seed43: train 5736 val 718 test 718


  fit 407.2s  nSV 4034


  gate OK 128 rows, min|dec| 4.376e-06 (36s)


  thinkpad    acc 0.6743  sink C_diminished_4 26.6%


  vivo        acc 0.9722  sink G_minor_4 4.1%


  flow        acc 0.9722  sink C_diminished_4 4.0%


  thinkpad-2  acc 0.3688  sink C_diminished_4 55.1%


  flow-2      acc 0.6417  sink C_diminished_4 26.6%


OK clean/seed43  test 1.0000  1.93 ms/clip


[clean/seed43 exit 0 in 499s]


clean/seed44: train 5736 val 718 test 718


  fit 262.8s  nSV 4034


  gate OK 128 rows, min|dec| 9.721e-07 (24s)


  thinkpad    acc 0.7688  sink C_diminished_4 19.6%


  vivo        acc 0.9785  sink G_minor_4 4.2%


  flow        acc 0.9833  sink F#_major_4 3.5%


  thinkpad-2  acc 0.5028  sink C_diminished_4 39.7%


  flow-2      acc 0.7556  sink C_diminished_4 14.9%


OK clean/seed44  test 1.0000  1.54 ms/clip


[clean/seed44 exit 0 in 343s]


clean/seed45: train 5736 val 718 test 718


  fit 281.8s  nSV 4017


  gate OK 128 rows, min|dec| 4.114e-06 (25s)


  thinkpad    acc 0.6618  sink C_diminished_4 29.0%


  vivo        acc 0.9688  sink G_minor_4 4.4%


  flow        acc 0.9785  sink C_diminished_4 4.1%


  thinkpad-2  acc 0.3993  sink C_diminished_4 54.7%


  flow-2      acc 0.6458  sink C_diminished_4 29.5%


OK clean/seed45  test 1.0000  2.67 ms/clip


[clean/seed45 exit 0 in 371s]


clean/seed46: train 5736 val 718 test 718


  fit 482.1s  nSV 4017


  gate OK 128 rows, min|dec| 7.288e-07 (21s)


  thinkpad    acc 0.6799  sink C_diminished_4 28.4%


  vivo        acc 0.9722  sink G_minor_4 4.4%


  flow        acc 0.9799  sink C_diminished_4 4.0%


  thinkpad-2  acc 0.3944  sink C_diminished_4 56.0%


  flow-2      acc 0.6611  sink C_diminished_4 28.1%


OK clean/seed46  test 1.0000  1.61 ms/clip


[clean/seed46 exit 0 in 581s]


clean-fixed/seed42: train 5736 val 718 test 718
